### PR TITLE
[SelectionDAG]: Deduce KnownNeverZero from SMIN and SMAX

### PR DESCRIPTION
**Pull request by AZero13**

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 1]
```diff
@@ -4017,6 +4015,16 @@ KnownBits SelectionDAG::computeKnownBits(SDValue Op, const APInt &DemandedElts,
       }
     }
 
+    // For SMIN, if CstHigh is negative we know the result will be
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4018]: I still don't understand why this part is needed. Does it affect codegen for any tests?

Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4018]: Is this tested?

Comment (AtariDreams) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4018]: Yes

Comment (AtariDreams) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4018]: Yes

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4018]: > Yes

Which test tests it? It wasn't obvious to me from the names.

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4018]: Don't "resolve conversations" initiated by someone else. Let them resolve it when they are satisifed.

Comment (AtariDreams) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4018]: smin_known_zero_2

it mirrors smax's logic

**Comment by jayfoad** [comment]
Maybe split this into two patches, one for `computeKnownBits` and one for `isKnownNeverZero`? Then it is easier to see what is being tested where.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp (hunk 1 above)
Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4018]: > smin_known_zero_2
> 
> it mirrors smax's logic

I don't see how that test could test something that uses ComputeNumSignBits.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 2]
```diff
@@ -5360,10 +5368,20 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
...
     // respectively we don't need the other to be known at all.
   case ISD::SMAX:
+    if (computeKnownBits(Op.getOperand(1), Depth + 1).isStrictlyPositive() ||
+        computeKnownBits(Op.getOperand(0), Depth + 1).isStrictlyPositive())
+      return true;
+    return isKnownNeverZero(Op.getOperand(1), Depth + 1) &&
+           isKnownNeverZero(Op.getOperand(0), Depth + 1);
   case ISD::SMIN:
+    if (computeKnownBits(Op.getOperand(1), Depth + 1).isNegative() ||
+        computeKnownBits(Op.getOperand(0), Depth + 1).isNegative())
```
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5381]: Both here and above, think you should cache the result of `computeKnownBits` so you can check if the `KnownBits` is trivially non-zero and potentially avoid 1/2 recursive calls through `isKnownNeverZero`.

Comment (AtariDreams) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5373]: Not sure how I would do that.

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5373]: Maybe pull out the computeKnownBits calls so we can also test:
```cpp
if (computeKnownBits(Op.getOperand(1), Depth + 1).isNonZero() &&
    computeKnownBits(Op.getOperand(0), Depth + 1).isNonZero())
```

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 3]
```diff
@@ -5360,10 +5360,42 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
...
+    KnownBits Op1 = computeKnownBits(Op.getOperand(1), Depth + 1);
+    if (Op1.isStrictlyPositive())
+      return true;
+
+    KnownBits Op0 = computeKnownBits(Op.getOperand(0), Depth + 1);
+
+    if (Op0.isStrictlyPositive())
+      return true;
+
+    if (computeKnownBits(Op.getOperand(1), Depth + 1).isNonZero() &&
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5375]: `Op1.isNonZero() && Op0.isNonZero()` don't call computeKnownBits again

Comment (AtariDreams) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp]: @topperc Addressed!

File: llvm/test/CodeGen/X86/known-never-zero.ll [hunk 4]
```diff
@@ -267,18 +267,31 @@ define i32 @smin_known_nonzero(i32 %xx, i32 %yy) {
...
+; CHECK-NEXT:    cmpl $-54, %edi
+; CHECK-NEXT:    movl $-54, %eax
+; CHECK-NEXT:    cmovll %edi, %eax
+; CHECK-NEXT:    rep bsfl %eax, %eax
+; CHECK-NEXT:    retq
+  %z = call i32 @llvm.smin.i32(i32 %x, i32 -54)
+  %r = call i32 @llvm.cttz.i32(i32 %z, i1 false)
+  ret i32 %r
+}
+
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/known-never-zero.ll:282]: Test vector cases

Comment (AtariDreams) [reply llvm/test/CodeGen/X86/known-never-zero.ll:282]: @llvm.cttz.i32 does not work on vectors/is not vectorizable

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 5]
```diff
@@ -5361,10 +5361,41 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
...
     // respectively we don't need the other to be known at all.
-  case ISD::SMAX:
-  case ISD::SMIN:
+  case ISD::SMAX: {
+    KnownBits Op1 = computeKnownBits(Op.getOperand(1), Depth + 1);
+    if (Op1.isStrictlyPositive())
+      return true;
+
+    KnownBits Op0 = computeKnownBits(Op.getOperand(0), Depth + 1);
+
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5372]: Uber-nit: remove this blank line?

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 6]
```diff
@@ -5361,10 +5361,41 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
...
+    return isKnownNeverZero(Op.getOperand(1), Depth + 1) &&
+           isKnownNeverZero(Op.getOperand(0), Depth + 1);
+  }
+  case ISD::SMIN: {
+    KnownBits Op1 = computeKnownBits(Op.getOperand(1), Depth + 1);
+    if (Op1.isNegative())
+      return true;
+
+    KnownBits Op0 = computeKnownBits(Op.getOperand(0), Depth + 1);
+
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5388]: And this one?

**Review by jayfoad — COMMENTED** [review]
Code LGTM but I haven't been involved too much in the review so I'll leave it for someone else to approve.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 7]
```diff
@@ -5361,10 +5361,41 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
...
+    if (Op0.isNegative())
+      return true;
+
+    if (computeKnownBits(Op.getOperand(1), Depth + 1).isNonZero() &&
+        computeKnownBits(Op.getOperand(0), Depth + 1).isNonZero())
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5393]: Use Op1 + Op0 KnownBits you've already calculated.

File: llvm/test/CodeGen/X86/known-never-zero.ll [hunk 8]
```diff
@@ -305,21 +360,77 @@ define i32 @smax_known_nonzero(i32 %xx, i32 %yy) {
...
-; CHECK-NEXT:    movl $32, %eax
-; CHECK-NEXT:    cmovnel %ecx, %eax
+; CHECK-NEXT:    rep bsfl %eax, %eax
 ; CHECK-NEXT:    retq
   %z = call i32 @llvm.smax.i32(i32 %x, i32 54)
   %r = call i32 @llvm.cttz.i32(i32 %z, i1 false)
   ret i32 %r
 }
 
+define <4 x i32> @smax_known_zero_vec(<4 x i32> %x, <4 x i32> %y) {
```
[the remark this answers is not in the document]
Comment (AtariDreams) [reply llvm/test/CodeGen/X86/known-never-zero.ll:376]: Perhaps because the instructions generated for vectors aren't using known bits?

**Review by arsenm — APPROVED** [review]

**Review by RKSimon — APPROVED** [review]
LGTM

EDIT - found an issue with the vector coverage

**Review by RKSimon — CHANGES_REQUESTED** [review]
You need better vector tests - the vector CTTZ expansion won't be making use of is never known zero

**Comment by AtariDreams** [comment]
> You need better vector tests - the vector CTTZ expansion won't be making use of is never known zero

So, what would then, if I may ask? Just to give me an idea so I could make a test for it

**Comment by RKSimon** [comment]
simplifySetCCWithCTPOP appears to use isKnownNeverZero for the `(ctpop x) eq/ne 1 --> (x & x-1) eq/ne 0` fold - maybe try that?

**Comment by AtariDreams** [comment]
@RKSimon Thank you!

**Review by RKSimon — CHANGES_REQUESTED** [review]
Please can you rebase - I've added better target coverage to known-never-zero.ll

**Comment by AtariDreams** [comment]
@RKSimon Done!

File: llvm/test/CodeGen/X86/known-never-zero.ll [hunk 9]
```diff
@@ -442,6 +445,88 @@ define i32 @smin_known_nonzero(i32 %xx, i32 %yy) {
...
+define i32 @smin_known_zero(i32 %x, i32 %y) {
+; CHECK-LABEL: smin_known_zero:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    cmpl $-54, %edi
+; CHECK-NEXT:    movl $-54, %eax
+; CHECK-NEXT:    cmovll %edi, %eax
+; CHECK-NEXT:    bsfl %eax, %ecx
+; CHECK-NEXT:    movl $32, %eax
+; CHECK-NEXT:    cmovnel %ecx, %eax
+; CHECK-NEXT:    retq
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/known-never-zero.ll:457]: Remove the CHECK lines, we use X86 and X64 in the file now.

**Comment by AtariDreams** [comment]
@RKSimon Done!

**Review by RKSimon — APPROVED** [review]

**Pull request merged**